cmd/cosign/cli/verify.go

```diff
@@ -218,43 +218,43 @@ You may specify either a key, a certificate or a kms reference to verify against
 
   # Verify a simple blob and message
-  cosign verify-blob --key cosign.pub --signature sig msg
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key cosign.pub --signature sig msg
 
   # Verify a simple blob with remote signature URL, both http and https schemes are supported
-  cosign verify-blob --key cosign.pub --signature http://host/my.sig
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key cosign.pub --signature http://host/my.sig
 
   # Verify a signature from an environment variable
-  cosign verify-blob --key cosign.pub --signature $sig msg
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key cosign.pub --signature $sig msg
 
   # verify a signature with public key provided by URL
-  cosign verify-blob --key https://host.for/<FILE> --signature $sig msg
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key https://host.for/<FILE> --signature $sig msg
 
   # Verify a signature against a payload from another process using process redirection
-  cosign verify-blob --key cosign.pub --signature $sig <(git rev-parse HEAD)
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key cosign.pub --signature $sig <(git rev-parse HEAD)
 
   # Verify a signature against Azure Key Vault
-  cosign verify-blob --key azurekms://[VAULT_NAME][VAULT_URI]/[KEY] --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key azurekms://[VAULT_NAME][VAULT_URI]/[KEY] --signature $sig <blob>
 
   # Verify a signature against AWS KMS
-  cosign verify-blob --key awskms://[ENDPOINT]/[ID/ALIAS/ARN] --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key awskms://[ENDPOINT]/[ID/ALIAS/ARN] --signature $sig <blob>
 
   # Verify a signature against Google Cloud KMS
-  cosign verify-blob --key gcpkms://projects/[PROJECT ID]/locations/[LOCATION]/keyRings/[KEYRING]/cryptoKeys/[KEY] --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key gcpkms://projects/[PROJECT ID]/locations/[LOCATION]/keyRings/[KEYRING]/cryptoKeys/[KEY] --signature $sig <blob>
 
   # Verify a signature against Hashicorp Vault
-  cosign verify-blob --key hashivault://[KEY] --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key hashivault://[KEY] --signature $sig <blob>
 
   # Verify a signature against GitLab with project name
-  cosign verify-blob --key gitlab://[OWNER]/[PROJECT_NAME]  --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key gitlab://[OWNER]/[PROJECT_NAME]  --signature $sig <blob>
 
   # Verify a signature against GitLab with project id
-  cosign verify-blob --key gitlab://[PROJECT_ID]  --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --key gitlab://[PROJECT_ID]  --signature $sig <blob>
 
   # Verify a signature against a certificate
-  cosign verify-blob --cert <cert> --signature $sig <blob>
+  COSIGN_EXPERIMENTAL=1 cosign verify-blob --cert <cert> --signature $sig <blob>
 `,
```

Comment:
@asraa Should we use the list you made of all possible flag combos (cert/key, experimental/no experimental, bundle, expired/unexpired cert) and have an example for each of these with more explanation?

Reply:
+1 

Further, I'd like to see some organization and explanation. Right now, the text is just a laundry list of "if X, then Y"

We should explain `unset COSIGN_EXPERIMENTAL` and `COSIGN_EXPERIMENTAL=1` separately (I don't 100% understand the current behavior so please correct me, indicated uncertainty with "?")

Here's a braindump, maybe should move into Google Doc (feel free to copy it)

# no experimental

Check the signature on a blob against the given key or certificate.

## Keys

blah

## Certificates

- `--certificate` *and* `--certificate-chain` example:

   The provided certificate (and all of the certificates in the chain(?) must be unexpired right now.

   **NOTE**: this just accepts whatever the last certificate of the chain is as your root. If you want to check that it matches a fixed root certificate, use keyless mode and `SIGSTORE_ROOT_FILE` (except we don't want people to do that? what do we tell these people?)


- `--certificate` no chain

   Checks that the certificate is signed by Fulcio (must be unexpired). Will that ever work?


## Bundles

- `--bundle`

    If `--bundle` is provided, use the given Rekor bundle (see ...) to check that the time of the *signature* was while the certificate (and its ancestors??) were valid (without `--bundle`, the certificate must be *currently* valid).

    Works for keys and certs both.

   What happens if the bundle has a conflicting chain?

- `--bundle` + certs, no chain

    If there's a bundle use that to get the chain up to Fulcio? 

# experimental

## Keys

Same as above, but we check that signatures are in Rekor (I think?)

If `--bundle` provided, check that against Rekor (offline).

##  Certs

- `--certificate` and `--certificate-chain`: Like the keyfull version, except instead of automatically trusting the certificate chain, we check that the root matches Fulcio's cert.

   If you'd like to BYO-root, set `SIGSTORE_ROOT_FILE` and we'll compare the root of the chain against that.

   We hit Rekor online.

- `--certificate`, no chain: this would only work if signed by the root directly, which probably won't be the case (maybe if you BYO-root?)

## Bundles 

- `--keys` Same as above, don't hit Rekor.

- `--certificate` and `--certificate-chain` and `--bundle`. Same as above,  don't hit rekor

    TODO: what if bundle has a conflicting cert/chain?

- `--certificate` and `--bundle`: Same, but use a chain from bundle. Fails if bundle has no chain. 

    TODO: what happens if the bundle has a conflicting cert?

- `--bundle`: Same, but use cert/chain from bundle. Fails if bundle has no cert/chain.

    TODO: if the bundle just provides a key, do we trust it?